File: src/main/java/rars/riscv/hardware/registerFiles/RegisterFileBase.java
package rars.riscv.hardware.registerFiles;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import rars.exceptions.SimulationException;
import rars.notices.RegisterAccessNotice;
import rars.riscv.hardware.registers.Register;
import rars.util.BinaryUtils;
import rars.util.ListenerDispatcher;

import java.util.ArrayList;

public abstract class RegisterFileBase {

    protected final @NotNull Register @NotNull [] registers;
    public final char registerNumberPrefix;
    
    private final @NotNull ArrayList<ListenerDispatcher<RegisterAccessNotice>.Handle> registerListeners;
    private final @NotNull ListenerDispatcher<@NotNull RegisterAccessNotice> registerChangeDispatcher;
    public final @NotNull ListenerDispatcher<@NotNull RegisterAccessNotice>.Hook registerChangeHook;

    protected RegisterFileBase(
        final char registerNumberPrefix,
        final @NotNull Register @NotNull [] registers
    ) {
        this.registerNumberPrefix = registerNumberPrefix;
        this.registers = registers;
        this.registerListeners = new ArrayList<>(registers.length);
        this.registerChangeDispatcher = new ListenerDispatcher<>();
        this.registerChangeHook = this.registerChangeDispatcher.getHook();
        for (final var register : this.registers) {
            final var handle = register.registerChangeHook.subscribe(
                this.registerChangeDispatcher::dispatch
            );
            this.registerListeners.add(handle);
        }
    }

    public final @Nullable Long updateRegisterByName(final @NotNull String registerName, final long newValue) throws
        SimulationException {
        final var register = this.getRegisterByName(registerName);
        if (register == null) {
            return null;
        }
        return this.updateRegister(register, newValue);
    }

    protected abstract int convertFromLong(final long value);

    public final long getLongValue(final @NotNull Register register) {
        return register.getValue();
    }

    public final int getIntValue(final @NotNull Register register) {
        return this.convertFromLong(register.getValue());
    }

    public final @Nullable Integer getIntValue(final @NotNull String registerName) {
        final var register = this.getRegisterByName(registerName);
        if (register == null) {
            return null;
        }
        return this.convertFromLong(register.getValue());
    }

    public final @Nullable Integer getIntValue(final int registerNumber) {
        final var register = this.getRegisterByNumber(registerNumber);
        if (register == null) {
            return null;
        }
        return this.convertFromLong(register.getValue());
    }

    public final @Nullable Long getLongValue(final @NotNull String registerName) {
        final var register = this.getRegisterByName(registerName);
        if (register == null) {
            return null;
        }
        return register.getValue();
    }

    public final @Nullable Long getLongValue(final int registerNumber) {
        final var register = this.getRegisterByNumber(registerNumber);
        if (register == null) {
            return null;
        }
        return this.getLongValue(register);
    }

    public final @Nullable Long updateRegisterByNumber(final int registerNumber, final long newValue) throws
        SimulationException {
        final var register = this.getRegisterByNumber(registerNumber);
        if (register == null) {
            return null;
        }
        return this.updateRegister(register, newValue);
    }

    public abstract long updateRegister(final @NotNull Register register, final long newValue) throws
        SimulationException;

    /**
     * Returns all the registers in the register file.
     *
     * @return An array of all the registers in the register file.
     */
    public @NotNull Register @NotNull [] getRegisters() {
        return this.registers;
    }

    public final @Nullable Register getRegisterByNumber(final int registerNumber) {
        for (final var register : this.registers) {
            if (register.number == registerNumber) {
                return register;
            }
        }
        return null;
    }

    public final @Nullable Register getRegisterByName(final @NotNull String name) {
        if (name.length() < 2) {
            return null;
        }

        // Handle a direct name
        for (final var register : this.registers) {
            if (register.name.equals(name)) {
                return register;
            }
        }
        // Handle prefix case
        if (name.charAt(0) == this.registerNumberPrefix) {
            if (name.charAt(1) == 0) { // Ensure that it is a normal decimal number
                if (name.length() > 2) {
                    return null;
                }
                return this.getRegisterByNumber(0);

            }

            final var integerNumber = BinaryUtils.stringToIntFast(name.substring(1));
            if (integerNumber == null) {
                return null;
            }
            return this.getRegisterByNumber(integerNumber);
        }
        return null;
    }

    public void resetRegisters() {
        for (final var register : this.registers) {
            register.resetValue();
        }
    }
}

File: src/main/java/rars/riscv/hardware/registerFiles/RegisterFileBase.kt
package rars.riscv.hardware.registerFiles

import arrow.core.Either
import arrow.core.raise.either
import rars.exceptions.SimulationError
import rars.notices.RegisterAccessNotice
import rars.riscv.hardware.registers.Register
import rars.util.Listener
import rars.util.translateToIntFast

abstract class RegisterFileBase protected constructor(
    private val registerNumberPrefix: Char,
    protected val myRegisters: Array<Register>
) {
    fun updateRegisterByName(registerName: String, newValue: Long): Either<SimulationError, Long?> = either {
        val register = this@RegisterFileBase.getRegisterByName(registerName)
        if (register == null) null
        else this@RegisterFileBase.updateRegister(register, newValue).bind()
    }

    /**
     * Returns all the registers in the register file.
     *
     * @return An array of all the registers in the register file.
     */
    open val registers: Array<Register> = myRegisters

    protected abstract fun convertFromLong(value: Long): Int

    fun getLongValue(register: Register): Long = register.getValue()

    fun getIntValue(register: Register): Int = this.convertFromLong(register.getValue())

    fun getIntValue(registerName: String): Int? {
        val register = this.getRegisterByName(registerName)
        if (register == null) {
            return null
        }
        return this.convertFromLong(register.getValue())
    }

    fun getIntValue(registerNumber: Int): Int? {
        val register = this.getRegisterByNumber(registerNumber)
        if (register == null) {
            return null
        }
        return this.convertFromLong(register.getValue())
    }

    fun getLongValue(registerName: String): Long? {
        val register = this.getRegisterByName(registerName)
        if (register == null) {
            return null
        }
        return register.getValue()
    }

    fun getLongValue(registerNumber: Int): Long? {
        val register = this.getRegisterByNumber(registerNumber)
        if (register == null) {
            return null
        }
        return this.getLongValue(register)
    }

    fun updateRegisterByNumber(registerNumber: Int, newValue: Long): Either<SimulationError, Long?> = either {
        val register = this@RegisterFileBase.getRegisterByNumber(registerNumber)
        if (register == null) null
        else this@RegisterFileBase.updateRegister(register, newValue).bind()
    }

    abstract fun updateRegister(register: Register, newValue: Long): Either<SimulationError, Long>

    fun getRegisterByNumber(registerNumber: Int): Register? {
        for (register in this.myRegisters) {
            if (register.number == registerNumber) {
                return register
            }
        }
        return null
    }

    fun getRegisterByName(name: String): Register? {
        if (name.length < 2) return null

        // Handle a direct name
        this.myRegisters.find { it.name == name }?.let { return it }

        // Handle prefix case
        return if (name[0] == this.registerNumberPrefix) {
            // Ensure that it is a normal decimal number
            if (name[1].code == 0) {
                if (name.length > 2) null
                else this.getRegisterByNumber(0)
            } else {
                val integerNumber = name.substring(1).translateToIntFast()
                integerNumber?.let { this.getRegisterByNumber(it) }
            }
        } else null
    }

    open fun resetRegisters() {
        for (register in this.myRegisters) {
            register.resetValue()
        }
    }

    fun addRegistersListener(listener: Listener<RegisterAccessNotice>) {
        for (register in this.myRegisters) {
            register.registerChangeHook.subscribe(listener)
        }
    }

    fun deleteRegistersListener(listener: Listener<RegisterAccessNotice>) {
        for (register in this.myRegisters) {
            register.registerChangeHook.unsubscribe(listener)
        }
    }
}
